Add merge style `fast-forward-only`

chrisnc opened the pull request:
With this option, it is possible to require a linear commit history with the following benefits over the next best option `Rebase+fast-forward`:
The original commits continue existing, with the original signatures continuing to stay valid instead of being rewritten, there is no merge commit, and reverting commits becomes easier. 

Closes #24906

chrisnc commented:
I'm missing the localization for the new UI, but otherwise I think it's close? This is my first contribution so I'm sure I've missed some other things, too. Reviews appreciated!

delvh commented on modules/structs/repo.go:
```diff
@@ -98,6 +98,7 @@ type Repository struct {
+	AllowFastForwardOnly          bool             `json:"allow_fast_forward_only_merge"`
 	AllowRebaseUpdate             bool             `json:"allow_rebase_update"`
 	DefaultDeleteBranchAfterMerge bool             `json:"default_delete_branch_after_merge"`
 	DefaultMergeStyle             string           `json:"default_merge_style"`
```
I think we need to change the usage of this variable too.
Or is it contained in `IsMergeStyleAllowed`?

chrisnc replied:
I think I've addressed the places where this field is used to account for `"fast-forward-only"`.

delvh commented on modules/structs/repo.go:
```diff
@@ -195,6 +196,8 @@ type EditRepoOption struct {
+	AllowFastForwardOnly *bool `json:"allow_fast_forward_only_merge,omitempty"`
 	// either `true` to allow mark pr as merged manually, or `false` to prevent it.
 	AllowManualMerge *bool `json:"allow_manual_merge,omitempty"`
 	// either `true` to enable AutodetectManualMerge, or `false` to prevent it. Note: In some special cases, misjudgments can occur.
```
7 lines below (I can't comment further down), the same problem might exist.

chrisnc replied:
Fixed.

delvh commented on services/pull/merge_ff_only.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+// doMergeStyleFastForwardOnly merges the tracking into the current HEAD - which is assumed to be staging branch (equal to the pr.BaseBranch)
+func doMergeStyleFastForwardOnly(ctx *mergeContext) error {
```
Please also add a test inside `tests/integration/pull_merge_test.go` for example.
Preferably, it covers at least the cases `fast forward is possible`, and `fast forward is not possible`

chrisnc replied:
I've started working on this using the other tests in that file as an example. As I was doing that, I started wondering if I should create a new `error` type for the case of a failed `--ff-only` merge because the PR head is not up-to-date, similar to `ErrMergeConflicts` and `ErrMergeUnrelatedHistories`? The other tests in that file use those for the failing cases, but maybe I shouldn't add my own error at this layer because it only applies to one merge style, whereas those other two should apply to any kind of merge? What do you think?

delvh replied:
I think `ErrMergeConflicts` is fine already as that's basically the reason why the merge is not working…

chrisnc replied:
Hm... the problem with that is that it's detected in a different way that won't apply to this case (`runMergeCommand` looks for `.git/MERGE_HEAD`). It's also misleading because an `--ff-only` merge will fail even if there are no conflicts, as long as the base has any commits that the head does not have.

delvh replied:
Yes, that was what I intended to imply: For a `ff-only` merge, the only possible merge conflicts are `branch is outdated` since there is no real merge.
Of course, I also have no problem with adding i.e. an `ErrBranchOutdated` or something like that.
If you want to go through the extra effort, fine with me.

chrisnc replied:
Having conflicts implies being outdated, but the reverse isn't true, so I think a new error makes sense here. I'll use a similar method as the `ErrMergeUnrelatedHistories` one, checking for a substring in `git`'s error output.

delvh replied:
Yep, I don't think you have any other choice here.

chrisnc replied:
New error is implemented along with integration tests for successful and failing `fast-forward-only` merges.

delvh commented on templates/repo/settings/options.tmpl:
```diff
@@ -517,6 +517,12 @@
+						<div class="field">
+							<div class="ui checkbox">
+								<input name="pulls_allow_fast_forward_only" type="checkbox" {{if or (not $pullRequestEnabled) ($prUnit.PullRequestsConfig.AllowFastForwardOnly)}}checked{{end}}>
+								<label>{{.locale.Tr "repo.pulls.fast_forward_only_merge_pull_request"}}</label>
```
At some point, we really need helper texts for these options.
They are tech jargon, and nowhere explained inside Gitea…
However, I don't know if that needs to be done in this PR.

chrisnc replied:
Created #28970.

Separately, on this line and a few other places I noticed I was missing the `ctx` before `.locale.Tr`. My original implementation of this predates that, and they didn't all show up as conflicts when I rebased if they were spaced out enough. I think they're all fixed now.

delvh reviewed (commented):
Found yet another place:
https://docs.gitea.com/next/administration/config-cheat-sheet#repository---pull-request-repositorypull-request `DEFAULT_MERGE_STYLE`.

delvh commented:
Oh, by the way:
You don't need to force push your branch.
We'll squash merge it anyway, so these other commits won't be added to the history, and force pushes make reviewing painful.

wxiaoguang commented on templates/repo/issue/view_content/pull.tmpl:
```diff
@@ -268,6 +268,13 @@
+								{
+									'name': 'fast-forward-only',
+									'allowed': {{and $prUnit.PullRequestsConfig.AllowFastForwardOnly (eq .Issue.PullRequest.CommitsBehind 0)}},
+									'textDoMerge': {{ctx.locale.Tr "repo.pulls.fast_forward_only_merge_pull_request"}},
```
`ctx.Locale` but not `ctx.locale`

chrisnc replied:
Fixed.

chrisnc commented:
Addressed the feedback so far, and the integration tests I added are passing. When I test this locally with a repository configured with only `fast-forward-only` merges allowed, I see the "This branch can be merged automatically" message for an outdated branch, which seems wrong, so I'll look into how to address that...

delvh commented:
> the integration tests I added are passing

CI disagrees, and says the routes haven't been found.

chrisnc commented:
> CI disagrees, and says the routes haven't been found.

Indeed. I ran `make integrations.sqlite.test` and it passed. Will fix the other errors.

chrisnc commented:
I think this is ready to go other than translations. Should I make an attempt at those or is there some other process for doing that?

delvh commented:
Ah, forgot to approve. Sorry.
No, they will be added automatically from Crowdin once the PR is merged.
